src/codegen/shuffle_bytes.rs

```diff
@@ -160,32 +162,84 @@ macro_rules! impl_shuffle_bytes {
+        cfg_if! {
+            if #[cfg(all(any(target_arch = "x86", target_arch = "x86_64"),
+                         target_feature = "avx"))] {
+                impl ShuffleBytes for u32x4 {
+                    #[inline]
+                    fn shuffle_bytes(self, indices: Self) -> Self {
+                        #[cfg(target_arch = "x86")]
+                        use arch::x86::{_mm_permutevar_ps};
+                        #[cfg(target_arch = "x86_64")]
+                        use arch::x86_64::{_mm_permutevar_ps};
+
+                        unsafe {
+                            mem::transmute(_mm_permutevar_ps(
+                                mem::transmute(self.0), mem::transmute(indices.0))
+                            )
+                        }
+                    }
+                }
+            } else {
+                impl ShuffleBytes for u32x4 {
+                    #[inline]
+                    fn shuffle_bytes(self, indices: Self) -> Self {
+                        let indices: u8x4 = (indices * 4).cast();
+                        let indices: u8x16 = shuffle!(indices, [0, 0, 0, 0, 1, 1, 1, 1,
+                                                                2, 2, 2, 2, 3, 3, 3, 3]);
+                        let v = u8x16::new(0, 1, 2, 3, 0, 1, 2, 3, 0, 1, 2, 3, 0, 1, 2, 3);
+                        let indices = indices + v;
+                        unsafe {
+                            let s: u8x16 = mem::transmute(self);
+                            mem::transmute(s.shuffle_bytes(indices))
+                        }
+                    }
                 }
             }
         }
     };
     (u64x2) => {
-        impl ShuffleBytes for u64x2 {
-            #[inline]
-            fn shuffle_bytes(self, indices: Self) -> Self {
-                let indices: u8x2 = (indices * 8).cast();
-                let indices: u8x16 = shuffle!(indices, [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1]);
-                let v = u8x16::new(0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7);
-                let indices = indices + v;
-                unsafe {
-                    let s: u8x16 = mem::transmute(self);
-                    mem::transmute(s.shuffle_bytes(indices))
+        cfg_if! {
+            if #[cfg(all(any(target_arch = "x86", target_arch = "x86_64"),
+                         target_feature = "avx"))] {
+                impl ShuffleBytes for u64x2 {
+                    #[inline]
+                    fn shuffle_bytes(self, indices: Self) -> Self {
+                        #[cfg(target_arch = "x86")]
+                        use arch::x86::{_mm_permutevar_pd};
+                        #[cfg(target_arch = "x86_64")]
+                        use arch::x86_64::{_mm_permutevar_pd};
+
+                        unsafe {
+                            mem::transmute(_mm_permutevar_pd(
+                                mem::transmute(self), mem::transmute(indices))
+                            )
+                        }
```

Comment:
Here `_mm_permutevar_pd` is used to permute the `u64x2` vector, but `shuffle_bytes` produces incorrect results for `(0, 1).shuffle_bytes([0, 1])`: it should produce (0, 1), but it produces (0, 0). 

The assembly generated for it is:

```asm
packed_simd::u64x2::shuffle_bytes:
 push    rbp
 mov     rbp, rsp
 vmovapd xmm0, xmmword, ptr, [rsi]
 vpermilpd xmm0, xmm0, xmmword, ptr, [rdx]
 vmovapd xmmword, ptr, [rdi], xmm0
 mov     rax, rdi
 pop     rbp
 ret
```

No idea what's going on here.

Reply:
cc @lemaitre

Reply:
Found it, [`_mm_permutevar_pd`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#expand=5236,2317,3233,5155,5206,2298,5482,4754,4754,3614,318,4422,4440,4449,3612,5241,2817,3064,5117,5019,5057,5105,4881,4437,4425,4425,1821,2969,5644,3262,1246,659,1,2786,1155,4654,4902,5084,5114,5084,2288,5028,4999,5028,4999,5028,4937,3155,3090,1049,4387,4401,4404,4422,575,5132,4422,2290,1821,4911,4902,659,659,659,4387,4397,4412,659,5114,1244,659,2977,1821,4654,1821,1821,3775,5723,1244,659,659,1244,1247,1246,3746,3253,659,3867,3231,932,3678,3678,3761,4795,4748,4754,585,5626,5626,97,243,243,91,659,4742,4813,4751,4813,659,3063,659,572,572,4920,221,572,4881,4872,572,4902,4872,3714,3746,3746,3714,3724,3762,3737,3717,3740,3714,3762,3746,221,4751,3714,4751,5084,3311,4092,3311,3311,3303,109,4725,5723,4751,3657,3660,3914,3915,3657,3657,3657,3657,3657,3657,3775,659,3714,79,1814,3657,3714,3660,2077,3657,3660,3657,3660,3657,3746,3737,3746,3762,3746,3746,3746,3737,4728,4729,97,88,151,152,106,97,3657,3657,3746,3737,3746,4195,3253,4751,3746,5084,3714,3939,3939,4798,3746,4813,4751,4813,4751,4742,4751,4792,4813,4754,4751,4753,4754,4792,3657,3657,3746,3657,1523,4742,5635,5670,4157,1523,406,4792,4813,4798,5093,3714,3703,3746,3737,3737,3936,4751,3936,4043,4751,4043,4090,1005,3549,432,406,432,5542,5545,765,432,432,406,432,3611,3507,3474,3567,4813,406,432,406,432,432,5653,5569,432,406,3378,406,673,243,4751,5132,575,5456,3611,5456,5455,5456,5456,5481,5487,5456,5455,5486,5486,891,4064,4043,1162,1129,5543,673,673,3549,3936,3311,3311,432,406,406,436,432,3311,3155,3644,1246,3936,3311,3909,3939,3936,3939,3930,3311,436,883,3660,436,427,413,427,413,427,3675,3737,4742,406,5123,4742,5723,4728,3103,4742,4620,3660,3657,3746,4701,4619,4670,4618,4613,4616,2288,3657,3660,4613,4696,4697,5093,5093,5123,4742,1815,1815,4742,5093,3906,88,3746,3737,4804,4795,3936,4795,4791,4742,4742,4751,4742,3936,4804,4751,4804,3814,3813,4792,4751,194,106,1564,5105,97,151,88,97,4798,97,4751,3906,4751,5093,4881,4872,3746,4751,97,4751,4742,3906,4742,3906,3936,3906,3897,3927,3927,3927&text=_mm_permutevar_pd) appears to use the **second** bit of each lane. Can't fathom why.

```rust
let x = __m128d::from_bits(u64x2::new(0, 1));
let y = __m128i::from_bits(u64x2::new(0b00, 0b10));
let z = unsafe { _mm_permutevar_pd(x, y) };
assert_eq!(u64x2::from_bits(z), u64x2::new(0, 1));

Reply:
> _mm_permutevar_pd appears to use the second bit of each lane

LOL. The saddest thing is that this doesn't surprise me. I think we can just map the (0, 1) indices that `shuffle_bytes` takes to `(0, 2)` by doing a multiply by `2` or a `shl` by `1`. I have no idea what's going to work better here. 

Reading the docs it appears that `_mm_permutevar_ps`  works as expected. There is a SELECT on the first two bits of each element of the `u32x4`. But that might be worth verifying.